torch/distributed/tensor/parallel/input_reshard.py

```diff
@@ -0,0 +1,133 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates
+import contextlib
+from functools import partial
+from typing import Any, Iterator, Optional, Tuple
+
+import torch
+from torch.distributed._tensor import DeviceMesh, DTensor, Replicate, Shard
+
+__all__ = [
+    "input_reshard_wrapper",
+]
+
+
+class InputReshard(torch.nn.Module):
+    """
+    An nn.Module that wraps another nn.Module with input resharding.
+    This is a wrapper created dedicated for Tensor Parallel so that
+    replicate input can be sharded after forward to save memory.
+    To shard and restore the input, we use
+    `torch.autograd.graph.saved_tensors_hooks`.
+    """
+
+    def __init__(
+        self,
+        mod: torch.nn.Module,
+        tp_device_mesh: DeviceMesh,
+        input_reshard_dim: Optional[int] = None,
+    ):
+        super(InputReshard, self).__init__()
+        self.inner_module = mod
+        self.mesh = tp_device_mesh
+        self.input_reshard_dim = input_reshard_dim
+
+    def named_parameters(
```

Comment:
I want to mention that doing this can be brittle for some cases since now iterating over `named_modules()` and constructing parameter names manually differs from `named_parameters()`. This is not necessarily common, but it can happen.

For example, nanoGPT's original optimizer construction did this (see the 'before'):
https://github.com/karpathy/nanoGPT/commit/7fe4a099ad2a4654f96a51c0736ecf347149c34c

Reply:
I see. I borrowed this idea from FSDP's AC wrapper. Hmm after reading composable API's AC, I think we might be able to make it a pre-forward hook.. let me try.